File: iris-ld/iris-ld-jsonld/src/main/java/city/smartb/iris/jsonld/JsonLdObject.java
package city.smartb.iris.jsonld;

import city.smartb.iris.jsonld.reader.JsonField;
import city.smartb.iris.jsonld.jackson.JsonFieldJackson;
import city.smartb.iris.jsonld.reader.JsonFieldReader;
import com.github.jsonldjava.core.JsonLdConsts;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class JsonLdObject {

    public static final String JSON_LD_CONTEXT = "@context";
    public static final String JSON_LD_ID = "id";
    public static final String JSON_LD_TYPE = "type";

    protected final LinkedHashMap<String, Object> jsonLdObject;
    private final JsonFieldReader fieldReader;

    public JsonLdObject(Map<String, Object> jsonLdObject) {
        this(jsonLdObject, JsonFieldJackson::new);
    }

    public JsonLdObject(Map<String, Object> jsonLdObject, JsonFieldReader fieldReader) {
        this.jsonLdObject = new LinkedHashMap<>(jsonLdObject);
        this.fieldReader = fieldReader;
    }

    public JsonField get(String key) {
        return fieldReader.read(jsonLdObject, key);
    }

    public List<Object> getContext() {
        return fieldReader.read(jsonLdObject, JsonLdConsts.CONTEXT).asListObjects(Object.class);
    }

    public String getId() {
        String ldId = fieldReader.read(jsonLdObject, JsonLdConsts.ID).asString();
        if(ldId != null) {
            return ldId;
        }
        return fieldReader.read(jsonLdObject, JSON_LD_ID).asString();
    }

    public String getType() {
        String ldId = fieldReader.read(jsonLdObject, JsonLdConsts.TYPE).asString();
        if(ldId != null) {
            return ldId;
        }
        return fieldReader.read(jsonLdObject, JSON_LD_TYPE).asString();
    }

    public Map<String, Object> toJSON() {
        return jsonLdObject;
    }
}

File: iris-ld/iris-ld-jsonld/src/main/kotlin/city/smartb/iris/ld/jsonld/JsonLdObject.kt
package city.smartb.iris.ld.jsonld

import city.smartb.iris.ld.jsonld.jackson.JsonFieldReaderJackson
import city.smartb.iris.ld.jsonld.reader.JsonField
import city.smartb.iris.ld.jsonld.reader.JsonFieldReader
import city.smartb.iris.ld.jsonld.serializer.JsonLdObjectDeserializer
import city.smartb.iris.ld.jsonld.serializer.JsonLdObjectSerializer
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize

@JsonSerialize(using = JsonLdObjectSerializer::class)
@JsonDeserialize(using = JsonLdObjectDeserializer::class)
open class JsonLdObject {
    protected var jsonLd: LinkedHashMap<String, Any>
    private var fieldReader: JsonFieldReader

    companion object {
        const val JSON_LD_CONTEXT = "@context"
        const val JSON_LD_ID = "id"
        const val JSON_LD_TYPE = "type"
    }

    constructor(jsonLdObject: Map<String, Any>) {
        this.jsonLd = LinkedHashMap(jsonLdObject)
        this.fieldReader = JsonFieldReaderJackson()
    }

    constructor(jsonLdObject: Map<String, Any>, fieldReader: JsonFieldReader) {
        this.jsonLd = LinkedHashMap(jsonLdObject)
        this.fieldReader = fieldReader
    }

    operator fun get(key: String): JsonField {
        return fieldReader.read(jsonLd, key)
    }

    val context: List<Any>
        get() = fieldReader.read(jsonLd, JsonLdConsts.CONTEXT).asListObjects(Any::class.java)

    open val id: String?
        get() {
            val ldId = fieldReader.read(jsonLd, JSON_LD_ID).asString()
            return ldId
        }
    val type: String?
        get() {
            val ldId = fieldReader.read(jsonLd, JsonLdConsts.TYPE).asString()
            return ldId ?: fieldReader.read(jsonLd, JSON_LD_TYPE).asString()
        }

    open fun asJson(): Map<String, Any> {
        return jsonLd
    }
}
